github/branches/retrieve_github_data.py

```diff
@@ -0,0 +1,610 @@
+    node = branch_protection.nodes()
+    _add_protection_fields(node)
+
+    # we'll have to iterate on matching branches for this branch
+    # protection rule
+    # Currently, we avoid iteration by tuning the value to never require
+    # multiple pages. This technique is brittle, but may be sufficient.
+    ref = node.matching_refs(first=50)
+    ref.total_count()
+    ref.page_info.__fields__(end_cursor=True, has_next_page=True)
+    ref.nodes().__fields__(
+        name=True, prefix=True,
+    )
+
+    return op
+
+
+def create_rule_query():
+    """Create a query object for additional branch protection entries.
+
+    Used to fetch subsequent pages. End Cursor is passed as a variable
+    """
+    op = Operation(schema.Query)
+
+    node = op.branch_protection_rules.nodes(cursor="$LAST_CURSOR")
+    _add_protection_fields(node)
+    return op
+
+
+# TODO Should be able to produce iterator for lowest level data
+
+
+def get_nested_branch_data(endpoint, reponame):
+    owner, name = reponame.split("/", 1)
+    op = create_operation(owner, name)
+
+    logger.info("Downloading base information from %s", endpoint)
+
+    d = endpoint(op)
+    errors = d.get("errors")
+    if errors:
+        endpoint.report_download_errors(errors)
+        return RepoBranchProtections(name_with_owner=reponame)
+
+    repodata = (op + d).repository
+
+    def _more_to_do(cur_result, fake_new_page=False):
+        """Determine if we need another query.
+
+        There are two nested repeating elements in the query - if either
+        is not yet exhausted, we have to do another query
+        """
+        # for hacky testing
+        if fake_new_page:
+            cur_result.branch_protection_rules.nodes[
+                0
+            ].matching_refs.page_info.has_next_page = True
+
+        has_more_rules = cur_result.branch_protection_rules.page_info.has_next_page
+        has_more_refs = any(
+            x.matching_refs.page_info.has_next_page
+            for x in cur_result.branch_protection_rules.nodes
+        )
+        return has_more_rules or has_more_refs
+
+    # TODO determine better way to test
+    fake_next_page = False
+
+    # TODO ensure errors are reported out to pytest when invoked from there
+    while _more_to_do(repodata, fake_next_page):
+        fake_next_page = False
+        # Need to work from inside out.
+        # and need to iterate over every rule -- KIS & hope for YAGNI
+        more_refs = any(
+            x.matching_refs.page_info.has_next_page
+            for x in repodata.branch_protection_rules.nodes
+        )
+        if more_refs:
+            # # setup for more matching branches
+            # page_info = repodata.branch_protection_rules.nodes[0].matching_refs.page_info
+            # logger.info(
+            #     "Downloading extra matching nodes: branches=%s", page_info
+            # )
+            # op = create_rule_query()
+            # gql_vars = {"$LAST_CURSOR": repodata}
+            # pass
```

Comment:
Should this commented out code be removed?

Reply:
Yeah -- bad habit from long ago - I should figure a way to detect-before-review.

removed in next commit.